server,docgen: enhance the API documentation

knz opened the pull request:
Informs https://github.com/cockroachdb/docs/issues/8602

This commit makes three changes:

- it extends the API doc generator to recognize
  annotations to distinguish “reserved”, “alpha” and “public” APIs
  and report this level of support in the generated output.

  [Example full output here](https://github.com/knz/cockroach/blob/20201022-nodes/docs/generated/http/full.md)

- it makes it possible to annotate different levels of support
  for different fields inside a single API payload. This makes
  it possible to document that certain fields of public APIs
  are “reserved”.

- it marks all the APIs as “reserved” except for the following three:

  - `/health` which has been stabilized and is safely documentable.

    [See doc here](https://github.com/knz/cockroach/blob/20201022-nodes/docs/generated/http/full.md#health)

  - `/_status/nodes` ([generated doc](https://github.com/knz/cockroach/blob/20201022-nodes/docs/generated/http/full.md#nodes)) and `/_status/hotranges` ([generated doc](https://github.com/knz/cockroach/blob/20201022-nodes/docs/generated/http/full.md#hotranges)) are marked
    as “alpha” as requested by
    @piyush-singh, to serve as example of how we could
    document an API.

Note that some other API endpoints could probably be marked as “alpha”
but the upcoming release cycle will see the introduction of a new
namespace for public APIs, so we wish to not over-commit at this time.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55854)
<!-- Reviewable:end -->

taroface reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55854)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @dhartunian, @knz, @piyush-singh, and @taroface)

___
*[docs/generated/http/full.md, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGGkEsB_TUUJcBFo8I:-MKGGkEsB_TUUJcBFo8J:bmp7zex) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/full.md#L187)):*
> ```Markdown
> The underlying response type is something we're looking to get rid of.
> 
> Note: this is an “alpha” API endpoint. It is subject to change without
> ```

(Just going to comment on the generated docs instead of the code.)

I feel the status notes are taking up too much space, being in four separate places: the endpoint description, request and response parameter descriptions, and the parameter tables. Instead, can we do something like this:

This `full.md` is intended to be included as a snippet in `api-reference.md` in `docs`: https://github.com/cockroachdb/docs/blob/e96f2137028b94b556280de845e201721558fbc2/v20.2/api-reference.md

- I can add a section header to `api-reference.md` called "Support Status" that explains how reserved, alpha, and public statuses are defined.
- Replace `{{.SupportStatus}}` with: `[status](#support-status)` which will link to the "Support Status" section header in the doc
  - For the endpoint and request/response parameter descriptions (i.e., outside the table cells), prepend `{{.SupportStatus}}` with **Support status:** so that it reads `**Support status:** [status](#support-status)`
___
*[docs/generated/http/full.md, line 235 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGL4Wp9wWVF6-2FySG:-MKGL4Wp9wWVF6-2FySH:b-uyh018) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/full.md#L235)):*
> ```Markdown
> NodeStatus records the most recent values of metrics for a node.
> 
> Note: this is an “alpha” API payload. It is subject to change without
> ```

It seems potentially confusing to label the payload as "alpha" when some of the individual fields (below) are "reserved". 

Is it actually necessary to give each payload an umbrella support status when this is also being displayed for each field? I wonder if we could just exclude the payload support status wherever there is a table?
___
*[docs/generated/http/nodes-request.md, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGNeVk2Ibg8PXmagh4:-MKGNeVk2Ibg8PXmagh5:bjk4zrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/nodes-request.md#L9)):*
> ```Markdown
> and the KV layer.
> 
> Note: this is an “alpha” API payload. It is subject to change without
> ```

Since there are no request parameters, I don't think the reference to a "payload" makes sense. But might be moot based on my other comments.

Also, would it be ok for me to exclude this `nodes-request` snippet from the API doc for `_status/nodes`?


<!-- Sent from Reviewable.io -->

taroface reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55854)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @dhartunian, @knz, and @piyush-singh)

___
*[docs/generated/http/nodes-other.md, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGRwYn-5cazzYgSgfc:-MKGRwYoB9T2Ub1uXsPu:b232sgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/nodes-other.md#L25)):*
> ```Markdown
> | activity | [NodeStatus.ActivityEntry](#cockroach.server.status.statuspb.NodeStatus.ActivityEntry) | repeated | activity is a map of nodeIDs to network statistics from this node to other nodes. | Note: this is a “reserved” API field and should not be relied upon to build external tools. No guarantee is made about its availability and stability in external uses. |
> | total_system_memory | [int64](#int64) |  | total_system_memory is the total RAM available to the system (or, if detected, the memory available to the cgroup this process is in) in bytes. | Note: this is an “alpha” API field. It is subject to change without advance notice in a subsequent release. |
> | num_cpus | [int32](#int32) |  | num_cpus is the number of logical CPUs as reported by the operating system on the host where the 'cockroach' process is running. Note that this does not report the number of CPUs actually used by 'cockroach'; this parameter is controlled separately. | Note: this is an “alpha” API field. It is subject to change without advance notice in a subsequent release. |
> ```

Instances of 'cockroach' should be \`cockroach\`

I can't recall if we have a process for suggesting copy edits on the field descriptions. Should I just open a PR on `cockroach`?


<!-- Sent from Reviewable.io -->

knz commented:
> I can't recall if we have a process for suggesting copy edits on the field descriptions.

I can't recall either, probably not discussed yet

> Should I just open a PR on cockroach?

yes that would certainly work

knz committed:
server,docgen: enhance the API documentation

This commit makes three changes:

- it extends the API doc generator to recognize
  annotations to distinguish “reserved”, “alpha” and “public” APIs
  and report this level of support in the generated output.

- it makes it possible to annotate different levels of support
  for different fields inside a single API payload. This makes
  it possible to document that certain fields of public APIs
  are “reserved”.

- it marks all the APIs as “reserved” except for the following three:

  - `/health` which has been stabilized and is safely documentable.
  - `/_status/nodes` and `/_status/hotranges` are marked
    as “alpha” as requested by
    @piyush-singh, to serve as example of how we could
    document an API.

Note that some other API endpoints could probably be marked as “alpha”
but the upcoming release cycle will see the introduction of a new
namespace for public APIs, so we wish to not over-commit at this time.

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55854)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @dhartunian, @jlinder, @piyush-singh, and @taroface)

___
*[docs/generated/http/full.md, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGGkEsB_TUUJcBFo8I:-MNdJ6E_7AD6WwNa6FCq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/full.md#L187)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

(Just going to comment on the generated docs instead of the code.)

I feel the status notes are taking up too much space, being in four separate places: the endpoint description, request and response parameter descriptions, and the parameter tables. Instead, can we do something like this:

This `full.md` is intended to be included as a snippet in `api-reference.md` in `docs`: https://github.com/cockroachdb/docs/blob/e96f2137028b94b556280de845e201721558fbc2/v20.2/api-reference.md

- I can add a section header to `api-reference.md` called "Support Status" that explains how reserved, alpha, and public statuses are defined.
- Replace `{{.SupportStatus}}` with: `[status](#support-status)` which will link to the "Support Status" section header in the doc
  - For the endpoint and request/response parameter descriptions (i.e., outside the table cells), prepend `{{.SupportStatus}}` with **Support status:** so that it reads `**Support status:** [status](#support-status)`
</blockquote></details>

Done.
___
*[docs/generated/http/full.md, line 235 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGL4Wp9wWVF6-2FySG:-MNdK1_a8AiwMHgnnPgJ:bva2bnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/full.md#L235)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

It seems potentially confusing to label the payload as "alpha" when some of the individual fields (below) are "reserved". 

Is it actually necessary to give each payload an umbrella support status when this is also being displayed for each field? I wonder if we could just exclude the payload support status wherever there is a table?
</blockquote></details>

Yes I think I was mistaken to include the support level on the request and response types. Removed.
___
*[docs/generated/http/nodes-other.md, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGRwYn-5cazzYgSgfc:-MNdJ84I2dz8Ym3moAvN:b-r00zkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/nodes-other.md#L25)):*
Done.

> I can't recall if we have a process for suggesting copy edits on the field descriptions. Should I just open a PR on cockroach?

Yes that could work.
___
*[docs/generated/http/nodes-request.md, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MKGNeVk2Ibg8PXmagh4:-MNdJTTUCL8U0joKoxzF:b-m8m3b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4156e4e2ce8f56ce92612857b39294b33f0883b/docs/generated/http/nodes-request.md#L9)):*
> Since there are no request parameters, I don't think the reference to a "payload" makes sense. But might be moot based on my other comments.

Indeed.

> Also, would it be ok for me to exclude this nodes-request snippet from the API doc for _status/nodes?

I suppose so - for empty request payloads, there's not much to document indeed.


<!-- Sent from Reviewable.io -->

knz commented:
Modified as suggested, RFAL

knz commented:
@taroface did you want to give this a new look? I addressed your previous comments.

taroface reviewed (changes requested):
Thank you for the updates and reminder. Just one more comment!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-:-MQON2-k531PXDw_e0EU:b-n96il7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @dhartunian, @jlinder, @knz, @piyush-singh, and @taroface)

___
*[docs/generated/http/full.md, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MQOKKtj95Y-6jAewkA1:-MQOKKtj95Y-6jAewkA2:bg53e8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9da4023c839d5a4ab6cc22ea5c2bbfb8b8e4e5a/docs/generated/http/full.md#L49)):*
> ```Markdown
> 
> 
> Support status: [reserved](#support-status)
> ```

I think support status can also be removed from here for the special types, right?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55854)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @dhartunian, @jlinder, @piyush-singh, and @taroface)

___
*[docs/generated/http/full.md, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-MQOKKtj95Y-6jAewkA1:-MQXs2tLDZZ1HBfT4DGu:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9da4023c839d5a4ab6cc22ea5c2bbfb8b8e4e5a/docs/generated/http/full.md#L49)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

I think support status can also be removed from here for the special types, right?
</blockquote></details>

Good idea. Done.


<!-- Sent from Reviewable.io -->

taroface reviewed (approved):
LGTM, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55854#-:-MQmkyxr6_Qx2nJweTVX:b2c0zcb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @dhartunian, @jlinder, @piyush-singh, and @taroface)



<!-- Sent from Reviewable.io -->

knz commented:
thanks!

bors r=taroface
